tests/results.cpp

```diff
@@ -2765,3 +2765,159 @@ TEMPLATE_TEST_CASE("results: aggregate", ResultsFromTable, ResultsFromQuery, Res
+            {
+                ++binding_context_end_notify_calls;
+            }
+        };
+        r->m_binding_context.reset(new Context());
+        coordinator->on_change();
+        r->begin_transaction();
+        REQUIRE(binding_context_start_notify_calls == 1);
+        REQUIRE(binding_context_end_notify_calls == 1);
+        r->cancel_transaction();
+    }
+
+    SECTION("notify BindingContext if no callbacks have been registered - remote commit") {
+        static int binding_context_start_notify_calls = 0;
+        static int binding_context_end_notify_calls = 0;
+        struct Context : BindingContext {
+            void will_send_notifications() override
+            {
+                ++binding_context_start_notify_calls;
+            }
+
+            void did_send_notifications() override
+            {
+                ++binding_context_end_notify_calls;
+            }
+        };
+        r->m_binding_context.reset(new Context());
+        coordinator->on_change();
+        r->begin_transaction();
+        REQUIRE(binding_context_start_notify_calls == 1);
+        REQUIRE(binding_context_end_notify_calls == 1);
+        r->cancel_transaction();
+    }
+    
+    SECTION("notify BindingContext before and after sending notifications from local commits") {
+        static int binding_context_start_notify_calls = 0;
+        static int binding_context_end_notify_calls = 0;
+        static int notification_calls = 0;
+
+        // Register normal notification callbacks
+        Results results1(r, table->where().greater_equal(0, 0));
+        Results results2(r, table->where().less(0, 10));
+
+        auto token1 = results1.add_notification_callback([&](CollectionChangeSet, std::exception_ptr err) {
+            REQUIRE_FALSE(err);
+            ++notification_calls;
+        });
+
+        auto token2 = results2.add_notification_callback([&](CollectionChangeSet, std::exception_ptr err) {
+            REQUIRE_FALSE(err);
+            ++notification_calls;
+        });
+
+        struct Context : BindingContext {
+            void will_send_notifications() override
+            {
+                REQUIRE(notification_calls == 0);
+                REQUIRE(binding_context_end_notify_calls == 0);
+                ++binding_context_start_notify_calls;
+            }
+
+            void did_send_notifications() override
+            {
+                REQUIRE(notification_calls == 2);
+                REQUIRE(binding_context_start_notify_calls == 1);
+                ++binding_context_end_notify_calls;
+            }
+        };
+        r->m_binding_context.reset(new Context());
+        coordinator->on_change();
+        r->begin_transaction();
+        table->add_empty_row();
+        r->commit_transaction();
+        REQUIRE(binding_context_start_notify_calls == 1);
+        REQUIRE(binding_context_end_notify_calls == 1);
+    }
+
+    SECTION("notify BindingContext before and after sending notifications from remote commits") {
```

Comment:
There's a lot of duplicated code between these two tests which could be eliminated by just pulling it out of the SECTION().